### ISSUE TITLE
WebMvcAutoConfiguration.java 540 GzipResourceResolver

### ISSUE THREAD

**Issue by CantosSong** [spring-projects/spring-boot#9158]
if config like this:

```properties
spring.resources.chain.enabled=true
spring.resources.chain.gzipped=true
spring.resources.chain.strategy.fixed.enabled=true**
spring.resources.chain.strategy.fixed.paths=/**/**
spring.resources.chain.strategy.fixed.version=v1.0.0
```

It doesn't work，after gzip compression，response is garbled.
so,i would have to change the method itself.Is that right?
sample：

```java
private void configureResourceChain(ResourceProperties.Chain properties,
				ResourceChainRegistration chain) {
			Strategy strategy = properties.getStrategy();
			if (strategy.getFixed().isEnabled() || strategy.getContent().isEnabled()) {
				chain.addResolver(getVersionResourceResolver(strategy));
			}
			if (properties.isGzipped()) {
				chain.addResolver(new GzipResourceResolver());
			}
			if (properties.isHtmlApplicationCache()) {
				chain.addTransformer(new AppCacheManifestTransformer());
			}
		}

private void configureResourceChain(ResourceProperties.Chain properties,
				ResourceChainRegistration chain) {
			Strategy strategy = properties.getStrategy();
			if (properties.isGzipped()) {
				chain.addResolver(new GzipResourceResolver());
			}
			if (strategy.getFixed().isEnabled() || strategy.getContent().isEnabled()) {
				chain.addResolver(getVersionResourceResolver(strategy));
			}
			if (properties.isHtmlApplicationCache()) {
				chain.addTransformer(new AppCacheManifestTransformer());
			}
		}
```

**Comment by bclozel** [comment]
I don't understand the issue.
Can you provide a sample (an actual project on github) that reproduces the behavior you're seeing and explain what's the behavior you expected?

Just so you know, the `GzipResourceResolver` does not compress anything; it just tries to resolve existing, already compressed, resources.

**Comment by CantosSong** [comment]
Should be due to the impact of version control, resulting in '.gz' file response garbled

**Comment by bclozel** [comment]
If a `.gz` file variant is available, it can be served with the appropriate `"Content-Encoding"` response header. Whether or not the response is garbled and why, has to be demonstrated by your sample so we can track down the issue.

**Comment by CantosSong** [comment]
Wait a moment。

**Comment by CantosSong** [comment]
https://github.com/CantosSong/sample/tree/master

**Comment by CantosSong** [comment]
sample https://github.com/CantosSong/sample/tree/master

**Comment by CantosSong** [comment]
@bclozel https://github.com/CantosSong/sample/tree/master

**Comment by CantosSong** [comment]
@bclozel  it's not work. https://github.com/CantosSong/sample/tree/c3322297ef3e66a347da28dd762d5dd411c4f9b8
it's ok.
https://github.com/CantosSong/sample/tree/master

**Comment by bclozel** [comment]
This issue is actually a duplicate of [SPR-14264](https://jira.spring.io/browse/SPR-14264) - this has been fixed and is scheduled for 5.0 (Spring Boot 2.0). This is not scheduled for backport as there are breaking changes involved. The [dedicated commit explains the problem](https://github.com/spring-projects/spring-framework/commit/ccb3c44dbcd4ab4741aec7b366ed43efa2f90b44).

The problem you're seeing here is that the `"Content-Encoding"` response header is missing form the HTTP response. The response is "garbled" since the browser doesn't expect the response to be gzipped (it is!).

Your sample suggests switching the order of `ResourceResolvers`. This indeed partially solves the issue for you, since the missing `"Content-Encoding"` is now there, but the other header added by the `VersionResourceResolver`, `"ETag"`, is now missing.

As a temporary workaround, I think you could disable the gzip resolver and just enable gzip compression on your servlet container or http proxy/CDN.
